### PR TITLE
Add `skimage` requirement to `ImageDiff` and add `TestHarness` run exception checking

### LINKED ISSUE #19334: `ImageDiff` does not catch import error and does not require skimage
## Bug Description
If the package `skimage` is not available, `ImageDiff` will not fail gracefully. It will throw an exception and kill the test harness, and some cases still exit 0.

## Steps to Reproduce
https://civet.inl.gov/job/880395/ see step "Test heat_condution", in which `skimage` is not available:

```
runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/mooseutils/ImageDiffer.py", line 128, in __compare
    import skimage.metrics
ModuleNotFoundError: No module named 'skimage'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 419, in runJob
    self.run(job) # Hand execution over to derived scheduler
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/RunParallel.py", line 44, in run
    output = tester.processResults(tester.getMooseDir(), self.options, job_output)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/testers/ImageDiff.py", line 66, in processResults
    differ = ImageDiffer(gold, test, allowed=allowed)
  File "/ramdisk/civet/build_0/moose/python/mooseutils/ImageDiffer.py", line 45, in __init__
    self.__compare()
  File "/ramdisk/civet/build_0/moose/python/mooseutils/ImageDiffer.py", line 132, in __compare
    import skimage.measure
ModuleNotFoundError: No module named 'skimage'

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 260, in queueJobs
    self.__status_pool_jobs.add(self.status_pool.apply_async(self.jobStatus, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 266, in queueJobs
    self.__runner_pool_jobs.add(self.run_pool.apply_async(self.runJob, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 260, in queueJobs
    self.__status_pool_jobs.add(self.status_pool.apply_async(self.jobStatus, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 260, in queueJobs
    self.__status_pool_jobs.add(self.status_pool.apply_async(self.jobStatus, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 260, in queueJobs
    self.__status_pool_jobs.add(self.status_pool.apply_async(self.jobStatus, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 260, in queueJobs
    self.__status_pool_jobs.add(self.status_pool.apply_async(self.jobStatus, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 260, in queueJobs
    self.__status_pool_jobs.add(self.status_pool.apply_async(self.jobStatus, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running

runWorker Exception: Traceback (most recent call last):
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 445, in runJob
    self.queueJobs(Jobs, j_lock)
  File "/ramdisk/civet/build_0/moose/python/TestHarness/schedulers/Scheduler.py", line 260, in queueJobs
    self.__status_pool_jobs.add(self.status_pool.apply_async(self.jobStatus, (job, Jobs, j_lock)))
  File "/data/spack/opt/spack/linux-centos8-zen2/gcc-9.3.0/python-3.7.8-p6j5c4qwjwwbhpelfbpvia2invtaa3ca/lib/python3.7/multiprocessing/pool.py", line 362, in apply_async
    raise ValueError("Pool not running")
ValueError: Pool not running


Successfully tested moose.

//: Singularity execution completed with return code 0
```

## Impact
The `TestHarness` can report positive results even though it failed.

### LINKED ISSUE #19338: `TestHarness` should catch python exceptions in a run and fail gracefully
## Reason
If python exceptions are encountered while running a `Tester`, the behavior is not well defined. Typically, the exception will end the `TestHarness` and possibly return with code 0.

## Design
Catch exceptions when running a `Tester` and set the `Tester` status to failed.

## Impact
More versatile testing. Catches errors and reports them appropriately.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #19334 and closes #19338

### CODE REVIEW

**Comment by loganharbour** [comment]
FYI @recuero

**Review by milljm — DISMISSED** [review]
👍

**Comment by permcody** [comment]
This looks reasonable to me, but I feel like the right design would have this kind of checking at either the Tester level (might be difficult) or the Scheduler level.

**Review by permcody — DISMISSED** [review]
If this passes, it's better than what we have now. I would recommend merging. I think there is still room for improvement in terms of where these checks occur in the abstraction hierarchy.

**Comment by loganharbour** [comment]
> This looks reasonable to me, but I feel like the right design would have this kind of checking at either the Tester level (might be difficult) or the Scheduler level.

I thought about this, but with all of the overridden methods it makes sense to also have a check where everything is called in case something is missed

File: python/TestHarness/schedulers/RunParallel.py [hunk 1]
```diff
@@ -25,46 +27,52 @@ def __init__(self, harness, params):
...
+                    if redirected_output == '':
+                        tester.setStatus(tester.fail, 'FILE TIMEOUT')
+                        output += '\n' + "#"*80 + '\nTester failed, reason: ' + tester.getStatusMessage() + '\n'
+
+                self.setSuccessfulMessage(tester)
+            else:
+                output += '\n' + "#"*80 + '\nTester failed, reason: ' + tester.getStatusMessage() + '\n'
+        except Exception as e:
+            output += 'Python exception encountered:\n\n' + traceback.format_exc()
+            tester.setStatus(tester.error, 'PYTHON EXCEPTION')
```
Comment (milljm) [inline python/TestHarness/schedulers/RunParallel.py:71]: 👍

**Review by milljm — APPROVED** [review]
looks good to me!

**Pull request merged**